bpf/process/bpf_rate.h

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#ifndef __RATE_H__
+#define __RATE_H__
+
+#include "bpf_tracing.h"
+#include "bpf_helpers.h"
+
+struct cgroup_rate_key {
+	__u64 id;
+};
+
+struct cgroup_rate_value {
+	__u64 curr;
+	__u64 prev;
+	__u64 time;
+	__u64 rate;
+	__u64 throttled;
+};
+
+struct cgroup_rate_options {
+	__u64 events;
+	__u64 interval;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+	__uint(max_entries, 32768);
+	__type(key, struct cgroup_rate_key);
+	__type(value, struct cgroup_rate_value);
+} cgroup_rate_map SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct cgroup_rate_options);
+} cgroup_rate_options_map SEC(".maps");
+
+#define NS (1000ULL * 1000ULL * 1000UL)
+
+static inline __attribute__((always_inline)) bool
+cgroup_rate(void *ctx, struct msg_k8s *kube, __u64 time)
+{
+	struct cgroup_rate_options *opt;
+	struct cgroup_rate_key key = {
+		.id = kube->cgrpid,
+	};
+	struct cgroup_rate_value *val;
+	__u64 delta, interval, slide;
+	__u32 zero = 0;
+
+	opt = map_lookup_elem(&cgroup_rate_options_map, &zero);
+	if (!opt)
+		return true;
+
+	interval = opt->interval;
+	if (!interval)
+		return true;
+
+	val = map_lookup_elem(&cgroup_rate_map, &key);
+	if (!val) {
+		struct cgroup_rate_value new_value = {
+			.time = (time / interval) * interval,
+			.curr = 1,
+		};
+
+		map_update_elem(&cgroup_rate_map, &key, &new_value, 0);
+		return true;
+	}
+
+	/*
+	 * We split the time in interval windows and keep track of events
+	 * of events count in current (val->curr) and previous (val->prev)
+	 * intervals.
+	 */
+
+	delta = time - val->time;
+	if (delta > interval) {
+		if (delta > 2 * interval) {
+			val->prev = 0;
+			val->time = (time / interval) * interval;
+		} else {
+			val->prev = val->curr;
+			val->time += interval;
+		}
+		val->curr = 0;
+	}
+
+	val->curr++;
+
+	/*
+	 * We compute the size of the slide window in previous interval and
+	 * based on that we compute partial amount of events from previous
+	 * interval window. Then we add current interval count and we have
+	 * rate value.
+	 *
+	 *                       val->time
+	 *                       |
+	 *   <--- interval ----->|<--- interval ----->|
+	 *                       |
+	 *    val->prev          | val->curr
+	 *   |-------------------|-----------
+	 *         val->rate
+	 *        |-------------------|
+	 *                            time
+	 */
+
+	slide = interval - (time - val->time);
+	val->rate = (slide * val->prev) / interval + val->curr;
```

Comment:
This calculates the rate of events per window but it assumes that the events in the prev window are evenly distributed. I think a better calculation would be:
val->rate = (val->prev + val->curr) / ((time - val->time) + interval)
[ extra brackets added for clarity ].
This calculation sums the previous window and the partial current window and then divides by the amount of time that passed since the beginning of the previous window. It does mean that the size of window that events are averaged over changes depending on the time and varies between a fraction over 1 interval, up to 2 intervals, and averages 1.5 intervals. Therefore if you set opt->interval to 2/3 the interval specified, then the window that events are averaged over will vary between 2/3 of the window specified and 4/3 of the window specified, averaging on exactly the window specified.
This approach doesn't care whether the events are evenly distributed or not.

Reply:
hum, you're right the current approach does not care whether events are evenly distributed,
but you suggestion is also approximation as you noted above.. I'll need to think some more on which one is worse/better ;-) at the moment both seem similar to me

Reply:
@kevsecurity I'm inclined in to keep the current computation for now, mainly due to:
  - both do approximation and are not accurate
  - I'm still not sure which one is more accurate, but we don't really need highly accurate rate, we aim to spot misbehaving cgroups that spike huge traffic that will cross the limit

I have already a follow up change to add task into the event, I'd try your way of rate computation as a follow up as well and see how it goes.. would that be ok?

Reply:
My suggestion is a better approximation because it doesn't assume an even distribution. Therefore I think it will always be more useful. Appreciate it might be unnecessary but I think it's worth a try, especially if you can compare the two approaches.

Reply:
A good property of the original approach is that it uses a precise count for the current period, so it can react better to spikes.

For example, let's say that our interval is 10, current time is 15, and that there were 10  events between 0 and 5, 10 events between 5 and 10, and 100 events between 10 and 15. The actual rate is 110, since there are 110 events between [5,15].

For the current approach is:
```
slide = 10 - (15-10) = 5
rate = 110
```

The suggested approach is:
```
(100 + 20) / (15 - 10 + 10) = 8
```

Which is 8*10 = 80 events on the interval.

Given the above, the first approach seems preferable to me.

Reply:
Okay, there are a few things that jump out.
1. The rates aren't directly comparable because (as explained in my comment) the window size needs adjusting to accommodate the different approach.
2. If you want to detect spikes smaller than the window, then you could make the window smaller and it would detect them just fine. By definition, it's the rate over a window size with all the events in the window being averaged over the length of the window. If you're interested in spikes smaller than this then a smaller window would be more appropriate as then the events will then be averaged over a shorter duration, which is more likely to detect a spike.
3. If the events were front-loaded, e.g. 100 between 0-5 and 0 between 5-15 then the rate between 5-15 should be 0, but the original would calculate it as rate = (5 * 100) / 10 + 0 = 50. So if we're just choosing specific examples, then we can easily choose numbers that favour one approach over the other. The bigger question is which wins out over a large number of simulations. I'll write a Monte Carlo simulation to test this, with the deciding factor being which approach is closest to the actual rate in each simulation, reported as a proportion of the total runs.

Reply:
Not sure I follow all the above. In my mind, it's not about how long the spike is but rather how fast you can detect it.

> So if we're just choosing specific examples, then we can easily choose numbers that favour one approach over the other

For this specific use-case, detecting spikes is more important than detecting sinks.

FWIW, I think that's an interesting discussion but I wouldn't want to block the PR on it. I would suggest we go with the existing approach and we can always change it if we find evidence that a better approach exists.

Reply:
I don't think there's evidence that one approach detects spikes quicker than the other, except for the singular example given. What we need to know is if that is representative or not. If the aim is to detect quicker, then make the window smaller and the threshold lower.

In the example I gave we would report a false spike when there is no spike. That's not about detecting dips, but about not detecting spikes when they aren't there (I assume we'd prefer to not throttle when we don't meet the requirements).

Reply:
> I don't think there's evidence that one approach detects spikes quicker than the other, except for the singular example given. What we need to know is if that is representative or not. If the aim is to detect quicker, then make the window smaller and the threshold lower.

I agree that it's not conclusive that one approach detects spikes quicker than the other. The example I used was just to show my intuition that this is the case because the rate of the current period is not smoothed. It would be interesting to see whether a counter example exists whether the spike is detected quicker with what you propose.

> In the example I gave we would report a false spike when there is no spike. That's not about detecting dips, but about not detecting spikes when they aren't there (I assume we'd prefer to not throttle when we don't meet the requirements).

This is an approximation so we can't be perfect no matter what we choose. For this use-case it would be preferable to throttle even if we don't reach the rate (false positive) than the other way round (true negative).

Reply:
sry for delay, github shows me replies with delay again.. 

I agree it's worth to try, still I think current way  has enough precision for what we need,
so I suggest I try your suggestion as a follow up and compare

Reply:
I ran a Monte Carlo simulation (I'll share code asynchronously so you can run it yourself and check the code) and my proposed way was more accurate in ~68% of cases.